### PR TITLE
feat(phases): wire real handler execution for `mcx phase run` (fixes #1381)

### LINKED ISSUE #1381: feat(phases): wire autonomous handler execution (mcx phase run without --dry-run)
Follow-up from #1299 / PR #1380.

Today `mcx phase run <name>` validates transitions and logs to `.mcx/transitions.jsonl`, but only `--dry-run` actually invokes the handler (via the logging proxy). The sprint phase scripts at \`.claude/phases/*.ts\` return discriminated \`{ action, ... }\` outputs that the orchestrator currently consumes by eye — to close the loop and get the "1 tool call per transition" payoff called for in #1286, real execution needs wiring.

Scope:
- \`mcx phase run <name>\` (no --dry-run) resolves + executes the handler with a real ctx (work-item-scoped state, MCP proxy backed by the daemon).
- Persist \`ctx.state\` writes to the per-work-item JSON blob at \`.mcx/state/<issueNumber>.json\` (amends #1290 per the #1299 spec).
- Surface the handler's structured return to stdout (JSON) so the orchestrator can pipe it.
- Respect \`runsOn\` and required-work-item invariants as gates.

Non-goals (separate follow-ups):
- \`onIdle\` re-entry / \`keepAlive\` semantics — the handlers today self-check state (e.g. \`if session_id exists, re-scan PR\`), which is the right primitive, but the orchestrator tick cadence is still manual.
- Quota pre-spawn gate in the manifest (\`gates.preSpawn\`).

Blocks the full #1299 dogfood target of "per-transition cost measurably lower (1 tool call vs 4-6)".

### PR DESCRIPTION
**Pull request by theshadow27**
## Summary
- `mcx phase run <target>` (no `--dry-run`) now executes the phase handler with a live `AliasContext`: daemon-backed state namespaced by work-item id, real MCP proxy, and `--work-item` resolved via IPC. Structured returns stream to stdout as JSON.
- `--no-execute` is the new escape hatch for orchestrators that want to record the transition separately from dispatch.
- `--arg key=val` and `--input <json>` carry handler inputs through, mirroring the dry-run runner.

## Adversarial-review resolution — two-phase transition log (#1407)

The first revision committed the transition to `.mcx/transitions.jsonl` **before** the handler ran. The reviewer flagged that this corrupted the audit log (logging never-happened transitions) and — more critically — broke re-entry: once `impl` was in history, a second `phase run impl` hit `RegressionError` because the real `.mcx.yaml` declares `impl.next = [triage]` (no self-loop).

Fix landed in `403f621d`:
- **`status: "attempted" | "committed"`** field on every transition log entry. Legacy entries (no `status`) are treated as `"committed"` for back-compat.
- **`"attempted"` entries** are appended before branch-guard / handler dispatch so every invocation leaves an audit trail (including crashes and wrong-branch attempts). They are **ignored** by graph-walk and regression checks.
- **`"committed"` entries** land only after the handler returns successfully. These drive `validateTransition`.
- **Idempotent self-loop rule** in `validateTransition`: `from === target && tail === target` is now a no-op re-entry, not a regression. Handlers are expected to be idempotent (self-check state, return `"in-flight"`), so successive `phase run X` calls must not trip `RegressionError`.
- **Pre-validation** runs before the `"attempted"` append, so unknown-phase / disallowed / un-forced-regression fail fast without wasting cycles on the handler.

### Trade-off
We deliberately emit TWO log lines per successful run (one `"attempted"`, one `"committed"`). The alternative — only committing, no attempt record — would lose audit evidence for branch-guard rejections and handler crashes, which was the author's original motivation for logging before execution. Two lines per run is cheap; lost audit evidence is not.

### Handler crash → retry
Handler throws → no `"committed"` entry. A retry re-enters cleanly because the transition log has only `"attempted"` entries for this work item, and those don't gate validation.

## Test plan
- [x] `bun typecheck`
- [x] `bun lint`
- [x] `bun test` (5070 pass, 0 fail)
- [x] `executePhase` tests cover:
  - real ctx dispatch (state persist + MCP call)
  - re-entry when state indicates in-flight (handler short-path)
  - branch-guard rejection on non-`runsOn` branches
  - missing work-item 404
  - handler-crash path — only `"attempted"` logged, no `"committed"` (updated invariant)
  - **new**: back-to-back `executePhase` calls on the same manifest/work-item don't trip `RegressionError` (idempotent self-loop)

## Notes
- Scope matches #1381: execute + emit JSON. Out of scope per the issue: `onIdle`/`keepAlive` re-entry cadence, pre-spawn quota gates. The CLI does **not** invoke spawn commands itself — that stays with the orchestrator.
- Existing tests that asserted the old transition-only behaviour were updated to pass `--no-execute`.
- Rebased on `origin/main` (picks up #1349, #1383, #1402).

## Remaining review nits (not addressed in this revision)
The 🟡/🔵 points from the review remain open — none are correctness blockers, filing follow-ups is appropriate:
- `Bun.spawnSync` in `defaultExecuteDeps.exec` has no `timeout:` option
- Silent JSON→string fallthrough on `--input` parse failure
- Asymmetric `cwd` forwarding on IPC methods
- `stateNamespace` using `:` delimiter

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by theshadow27** [comment]
## Adversarial Review (Delta)

### Changes Since Last Review

| # | Previous Issue | Status | Notes |
|---|----------------|--------|-------|
| 1 | 🔴 Transition committed before handler runs → `RegressionError` on re-entry | ✅ Fixed in `403f621d` | Two-phase log (`attempted`/`committed`) + `isCommitted` filter in `commitTransition`/pre-validate + idempotent self-loop rule in `validateTransition` (`phase-transition.ts:184-193`). Committed entries are the only ones gating future transitions. |
| 2 | 🟡 "Idempotent re-entry" test is a straw-man (pre-populated state, never ran twice) | ✅ Fixed in `403f621d` | New test `"back-to-back executePhase calls do not trip RegressionError (#1407)"` at `phase.spec.ts:1611` calls `executePhase` twice against the same manifest/work-item — second call returns `"in-flight"` and log contains 2 committed + 2 attempted entries. |
| 3 | 🟡 Handler-crash test asserts wrong invariant | ✅ Fixed in `403f621d` | Test renamed to `"handler crash leaves only an 'attempted' entry — no committed transition"` (`phase.spec.ts:1551`). Asserts `attempted.length === 1`, `committed.length === 0`, and that `"approved"` is NOT logged on crash. Retry path is covered by the back-to-back test above. |
| 4 | 🟡 `Bun.spawnSync` branch-guard has no timeout | ⏳ Not addressed | Author acknowledges in PR body as follow-up. Filing separately. |
| 5 | 🟡 Silent JSON→string fallthrough on `--input` | ⏳ Not addressed | Acknowledged as follow-up. |
| 6 | 🔵 Asymmetric `cwd` forwarding on IPC | ⏳ Not addressed | Acknowledged as follow-up. |
| 7 | 🔵 `stateNamespace = workitem:${id}` delimiter fragility | ⏳ Not addressed | Acknowledged as follow-up. |
| 8 | 🔵 DI fragility after `d.exit(1)` in catch | ⏳ Not addressed | Not raised in PR body; minor. |

### Verification
- Ran `bun test packages/command/src/commands/phase.spec.ts` locally against `403f621d`: **112 pass, 0 fail** (1.79s).
- `validateTransition` self-loop rule correctly requires `from !== null && from === target && history[last] === target` — won't mask genuine `A → B → A` regressions (those fail because `from=B`, not `A`).
- `commitTransition` and `executePhase` pre-validate both filter history with `isCommitted` so `attempted` entries never gate.
- `readAllTransitions` (used by `mcx phase log` at `phase.ts:670`) intentionally returns all entries for audit visibility — correct.

### New Issues

- 🔵 **Attempted-entry growth unbounded on retry loops** (`phase-transition.ts:309-339`). Every crashed handler dispatch leaves another `"attempted"` line. For long-running work items with flaky handlers, the jsonl will grow without compaction. Not a correctness issue (readers filter for display/gating), but worth a follow-up for log rotation or periodic compaction. Low priority.

- 🔵 **`appendAttempt` bypasses the transition lock** (`phase-transition.ts:315`, uses `appendTransitionLog` directly, not `withTransitionLock`). `appendFileSync` of a single short line is atomic on POSIX, so interleaved writes from concurrent `phase run` invocations won't corrupt the file — but this is a latent assumption. If attempt entries ever grow (e.g., embedded stack trace, large input hash), byte-level interleaving becomes possible. Document the invariant or take the lock.

- 🔵 **Docstring drift** (`phase.ts:950`). The updated docstring says "captures attempt evidence from ANY branch, including cases that branch-guard rejects or handlers crash" — but pre-validation (unknown phase, disallowed, un-forced regression) runs *before* `appendAttempt`, so those failures leave no audit entry. Minor inconsistency; either tighten the wording or move `appendAttempt` above pre-validate (the latter loses the fail-fast benefit the author called out).

### Suggested Automation

- Property test: for any manifest + any valid phase target, running `executePhase` N times in a row must either succeed or surface a domain-specific error — never `RegressionError`. Would have caught the original bug; guards against future state-machine regressions.
- (From round 1, still applicable) Lint rule: `Bun.spawnSync` in CLI hot paths without `timeout:`.

### Updated Verdict
✅ **Approve** — the 🔴 blocker from round 1 is fixed with the right semantics (two-phase log + idempotent self-loop), the new tests exercise both the handler-throws case and the real re-entry path, and all 112 spec tests pass. Remaining items are 🟡/🔵 follow-ups the author already acknowledged plus three new low-severity 🔵 notes worth filing but not blocking.

**Comment by theshadow27** [comment]
## QA Verification

**PR:** #1407 — feat(phases): wire real handler execution for `mcx phase run` (fixes #1381)
**Linked issue:** #1381

### Implementation Evidence
- `packages/core/src/phase-transition.ts`: `TransitionStatus` type + `isCommitted()` helper (line 36-56); `appendAttempt()` (line 317); `commitTransition` filters to committed-only history (line 471); idempotent self-loop rule in `validateTransition` (line 184-193)
- `packages/command/src/commands/phase.ts`: `cmdPhase` dispatch routes non-`--dry-run` / non-`--no-execute` to new `executePhase()` (line 712-713); `--no-execute` preserves old transition-only path (line 704-711); `parsePhaseExecuteArgs()` handles `--arg`/`--input` (line 878); `executePhase()` implements two-phase log with real ctx (line 973); `makeRealMcpProxy()` (line 945); branch-guard via `checkRunsOn` (line 1043); daemon work-item fetch (line 1064-1079); `createAliasState` + `createAliasCache` (line 1087-1102); structured JSON output to stdout (line 1145)

### Test Coverage
- `packages/command/src/commands/phase.spec.ts`: `parsePhaseExecuteArgs` suite — 5 unit tests for arg parsing edge cases (line 1267)
- `executePhase` suite — 6 integration tests covering: real ctx dispatch with state persist + MCP call; in-flight re-entry; branch-guard rejection; missing work item 404; handler crash → only `attempted` logged (no `committed`); back-to-back calls don't trip `RegressionError` (line 1355+)
- Existing tests updated to use `--no-execute` for transition-only behavior

### Test Results
```
5070 pass, 0 fail — bun test (full suite)
112 pass, 0 fail — packages/command/src/commands/phase.spec.ts
33 pass, 0 fail — packages/core/src/phase-transition.spec.ts
```

### CI Status
All three required checks pass (build, check, coverage) — https://github.com/theshadow27/mcp-cli/actions/runs/24412791876

### Remaining Work (non-blocking, acknowledged in PR body)
- `Bun.spawnSync` no `timeout:` in `defaultExecuteDeps.exec`
- Silent JSON→string fallthrough on `--input` parse failure
- Asymmetric `cwd` forwarding on IPC methods
- `stateNamespace` using `:` delimiter

### Open Inline Threads (blocking)

The 4 Copilot inline threads are unresolved and must be addressed before merge:

1. **phase.ts:863** — `exitCode ?? 0` masks killed/signal-terminated processes as success; `checkRunsOn` could pass when git was killed. Fix or dismiss with a follow-up issue.
2. **phase.ts (docstring)** — Numbered steps in `executePhase` JSDoc list branch-guard before commit, but implementation does it after. Doc is inconsistent. Fix the comment or dismiss.
3. **phase.ts:1098** — `stateNamespace = "phase:${target}:unbound"` is a stable key, not ephemeral. Comment says writes are "discarded" but they persist across unrelated unbound runs via daemon IPC. Either use a per-invocation namespace or a no-op stub, and update the comment.
4. **phase.ts:945** — `makeRealMcpProxy` duplicates proxy logic from `alias-runner.ts`. Not a correctness blocker; dismiss or file a follow-up.

### Outcome
**qa:fail** — Implementation and tests are correct; CI is green. Blocked by 4 open inline Copilot threads per QA policy. Action for implementation: resolve each thread (address the suggestion, dismiss with a comment, or file a follow-up issue and mark resolved).

File: packages/command/src/commands/phase.ts [hunk 1]
```diff
@@ -815,6 +831,274 @@ async function runPhase(argv: string[], d: PhaseInstallDeps): Promise<void> {
...
+  findGitRoot: (cwd: string) => string | null;
+  now: () => Date;
+}
+
+const defaultExecuteDeps: PhaseExecuteDeps = {
+  ipcCall,
+  exec: (cmd: string[]): ExecResult => {
+    const [bin, ...rest] = cmd;
+    const r = Bun.spawnSync([bin, ...rest], { stdout: "pipe", stderr: "pipe" });
+    return { stdout: new TextDecoder().decode(r.stdout), exitCode: r.exitCode ?? 0 };
```
[the remark this answers is not in the document]
Comment (theshadow27) [reply packages/command/src/commands/phase.ts]: Fixed in 988fe8e4. `exitCode` is now `r.exitCode ?? 1` — signal-terminated children surface as failure instead of being coerced to 0 (branch-guard would otherwise wave through a git that never ran).

File: packages/command/src/commands/phase.ts [hunk 2]
```diff
@@ -815,6 +831,274 @@ async function runPhase(argv: string[], d: PhaseInstallDeps): Promise<void> {
...
+ * Flow:
+ *   1. Parse flags (transition + execution inputs)
+ *   2. Branch guard: refuse to run outside the manifest's `runsOn` branch
+ *   3. Commit the transition via `phaseRun` (same as --no-execute path)
+ *   4. Bundle the phase source
+ *   5. Fetch the work item from the daemon (if `--work-item` given)
+ *   6. Build a live AliasContext: real MCP proxy + daemon-backed state
+ *      (namespaced by the work-item id, so state writes persist per-item)
+ *   7. Execute the handler and emit its structured return on stdout
+ *
```
[the remark this answers is not in the document]
Comment (theshadow27) [reply packages/command/src/commands/phase.ts]: Fixed in 988fe8e4. Doc now reflects the actual order: pre-validate → append attempted → branch guard → bundle → fetch work item → build ctx → execute handler → commit transition (only on success).

File: packages/command/src/commands/phase.ts [hunk 3]
```diff
@@ -815,6 +831,274 @@ async function runPhase(argv: string[], d: PhaseInstallDeps): Promise<void> {
...
+      d.exit(1);
+    }
+  }
+
+  const repoRoot = ex.findGitRoot(cwd) ?? NO_REPO_ROOT;
+  // State is namespaced by work-item id so every phase touching the same
+  // item sees the same scratchpad (see sprint state declarations in
+  // .mcx.yaml). When no work item is bound, writes are discarded into an
+  // ephemeral namespace so the handler doesn't fault on missing state.
+  const stateNamespace = workItem ? `workitem:${workItem.id}` : `phase:${parsed.target}:unbound`;
```
[the remark this answers is not in the document]
Comment (theshadow27) [reply packages/command/src/commands/phase.ts]: Fixed in 988fe8e4. Comment corrected — `createAliasState` always persists through the daemon; when no work item is bound we just namespace into `phase:<target>:unbound` so writes don't collide with work-item scoped state. Not an in-memory discard.

File: packages/command/src/commands/phase.ts [hunk 4]
```diff
@@ -815,6 +831,274 @@ async function runPhase(argv: string[], d: PhaseInstallDeps): Promise<void> {
...
+              arguments: toolArgs ?? {},
+              cwd,
+            });
+            return extractContent(result);
+          };
+        },
+      });
+    },
+  });
+}
```
[the remark this answers is not in the document]
Comment (theshadow27) [reply packages/command/src/commands/phase.ts]: Fixed in 988fe8e4. Extracted the shared wiring into `@mcp-cli/core` as `createMcpProxy({ call, cwd })`. Both `alias-runner.ts` and `phase.ts` now consume it — no more drifting copies.

**Comment by theshadow27** [comment]
## QA Verification

**PR:** #1407 — feat(phases): wire real handler execution for `mcx phase run` (fixes #1381)
**Linked issue:** #1381

### Copilot Comment Resolution (primary task)

This QA round focuses on the 4 Copilot review comments flagged in the previous `qa:fail` and addressed in commit `988fe8e4`.

| Comment ID | Issue | Fix in 988fe8e4 | Status |
|---|---|---|---|
| 3081154383 | `exitCode ?? 0` masks signal-killed processes as success | Changed to `exitCode ?? 1` — `packages/command/src/commands/phase.ts:863` | ✅ Resolved |
| 3081154433 | Doc comment step order doesn't match implementation | Doc rewritten to match actual flow (pre-validate → attempted → branch guard → handler → commit) — `phase.ts:930-960` | ✅ Resolved |
| 3081154451 | Comment said "writes discarded into ephemeral namespace" but `createAliasState` always persists via daemon | Comment corrected: "unbound" lands in `phase:<target>:unbound` namespace in SQLite, not discarded — `phase.ts:1083-1087` | ✅ Resolved |
| 3081154480 | `makeRealMcpProxy` duplicated proxy logic from `alias-runner.ts` | Extracted into `packages/core/src/mcp-proxy.ts` as `createMcpProxy`. Both `alias-runner.ts` (line 30) and `phase.ts` (line 1090) now import from `@mcp-cli/core` | ✅ Resolved |

### Implementation Evidence
- `packages/core/src/mcp-proxy.ts` — new shared proxy factory, exported via `packages/core/src/index.ts`
- `packages/command/src/commands/phase.ts:863` — `exitCode ?? 1`
- `packages/command/src/commands/phase.ts:1081-1087` — corrected stateNamespace comment
- `packages/command/src/alias-runner.ts:30` — uses `createMcpProxy({ cwd: () => process.cwd() })` from core

### Test Coverage
- Existing `executePhase` spec suite covers: real ctx dispatch, re-entry, branch guard rejection, missing work-item 404, handler crash (only `attempted` logged), back-to-back idempotent self-loop

### Test Results
5059 pass, 0 fail across 210 files (local run)

### CI Status
In progress at time of QA — `check` and `coverage` jobs running. Typecheck passed locally; all steps up to daemon tests show green in CI.

### Remaining Work (non-blocking, noted in PR body)
- `Bun.spawnSync` in `defaultExecuteDeps.exec` has no `timeout:` option
- Silent JSON→string fallthrough on `--input` parse failure
- Asymmetric `cwd` forwarding on IPC methods
- `stateNamespace` using `:` delimiter

### Outcome
**qa:pass** — All 4 Copilot review comments resolved with evidence. Tests green. Implementation matches spec.
